### PR TITLE
Rework maps

### PR DESCRIPTION
**Pull request by fu5ha**
The main ideas:

- Remove dependence on `BTreeMap` everywhere and replace it with new `LinearMap` from `kollect` crate
- Support derive reflect for `std::HashMap` 🎉 (and the new `tame_containers::{LinearMap, OrderedMap, UnorderedMap}`)
- ~~Add test for static `TypeDescriptor` hashing so that we can make sure we don't break things in the future unintentionally~~ Postponed after further thought

### CODE REVIEW

**Comment by fu5ha** [comment]
Hm, new static hash test fails on CI which is concerning. Will need to investigate that more I guess..

File: crates/mirror-mirror-macros/src/derive_reflect/struct_named.rs
Comment (davidpdrsn) [inline crates/mirror-mirror-macros/src/derive_reflect/struct_named.rs:412]: A little unfortunate for the non-mutating case to require a box. Why is that?

File: crates/mirror-mirror/src/lib.rs
Comment (davidpdrsn) [inline crates/mirror-mirror/src/lib.rs:199]: Wanna mention the new `simple_type_name` feature here?

**Review by davidpdrsn — COMMENTED** [review]
Awesome! Just a couple of small questions.

Wanna update the changelog as well?

File: crates/mirror-mirror-macros/src/derive_reflect/struct_named.rs
Comment (fu5ha) [reply crates/mirror-mirror-macros/src/derive_reflect/struct_named.rs:412]: I believe this will actually be faster than the previous version. Box is a little unfortunate indeed but the reason is I wanted to be able to specialize the iterator type per implementation. Previously it didn't allocate but on each `next()` call there was quite a lot to do. Now we box, but each call to `next()` should be trivial. Which I think will actually win out.

File: crates/mirror-mirror/src/lib.rs
Comment (fu5ha) [reply crates/mirror-mirror/src/lib.rs:199]: Ah, yeah

File: crates/tame-containers/src/tests/ordered_map.rs
Comment (fu5ha) [reply crates/tame-containers/src/tests/ordered_map.rs:1]: for the future :P

**Comment by fu5ha** [comment]
Don't merge yet -- moving `tame-containers` to other repo.

**Comment by fu5ha** [comment]
Okay ready for re-review after published `kollect` 0.1.0. Moved to all `LinearMap` as briefly suggested on slack

**Review by davidpdrsn — APPROVED** [review]
🚀

**Comment by fu5ha** [comment]
Alright, merging... I'm still concerned about what this means for serde backcompat so we'll have to revisit that before merging main/releasing 0.2

**Pull request merged**
